spec/functional/run_lock_spec.rb

```diff
@@ -267,20 +311,120 @@ def results
-        run_lock.save_pid
-        sleep 2
-        exit! 1
+  end
+
+  class ClientProcess
+    def initialize(example, name)
+      @example = example
+      @name = name
+      @read_from_process, @write_to_tests = IO.pipe
+      @read_from_tests, @write_to_process = IO.pipe
+    end
+
+    attr_reader :example
+    attr_reader :name
+    attr_reader :pid
+
+    def event(event)
+      # Let the caller know what event we've reached
+      write_to_tests.puts(event)
+
+      # Block until the client tells us where to stop
+      if !@run_to_event || event == @run_to_event
+        @run_to_event = read_from_tests.gets.strip
       end
+    end
 
-      wait_on_lock
-      sleep 0.5 # Possible race condition on Solaris which pid is observed as 0
-      expect(File.read(lockfile)).to eq(p1.to_s)
+    def last_event
+      while true
+        event = readline_nonblock(read_from_process)
+        break if event.nil?
+        @last_event = event.strip
+      end
+      @last_event
+    end
 
-      Process.waitpid2(p1)
+    def run_to(event, &background_block)
+      # Start the process if it's not started
+      start if !pid
+
+      # Tell the process what to stop at (also means it can go)
+      write_to_process.puts event
+
+      # Run the background block
+      background_block.call if background_block
+
+      # Wait until it gets there
+      Timeout::timeout(CLIENT_PROCESS_TIMEOUT) do
+        until last_event == event
+          @last_event = read_from_process.gets.strip
+        end
+      end
+    end
+
+    def run_to_completion
+      # Start the process if it's not started
+      start if !pid
+
+      # Tell the process to stop at nothing (no blocking)
+      @write_to_process.puts "nothing"
+
+      # Wait for the process to exit
+      wait_for_exit
+    end
+
+    def wait_for_exit
+      Timeout::timeout(CLIENT_PROCESS_TIMEOUT) do
+        Process.wait(pid) if pid
+      end
+    end
+
+    def stop
+      if pid
+        begin
+          Process.kill(:KILL, pid)
+          Timeout::timeout(CLIENT_PROCESS_TIMEOUT) do
+            Process.waitpid2(pid)
+          end
+        # Process not found is perfectly fine when we're trying to kill a process :)
+        rescue Errno::ESRCH
+        end
+      end
+    end
+
+    private
+
+    attr_reader :read_from_process
+    attr_reader :write_to_tests
+    attr_reader :read_from_tests
+    attr_reader :write_to_process
+
+    def start
+      @pid = fork do
+        Timeout::timeout(CLIENT_PROCESS_TIMEOUT) do
+          run_lock = Chef::RunLock.new(example.lockfile)
+          event("started")
+          run_lock.acquire
+          event("acquired lock")
+          run_lock.save_pid
+          event("saved pid")
+          exit!(0)
+        end
+      end
     end
 
+    def readline_nonblock(fd)
+      buffer = ""
+      buffer << fd.read_nonblock(1) while buffer[-1] != "\n"
+
+      buffer
+    #rescue IO::EAGAINUnreadable
+    rescue IO::WaitReadable
+      unless buffer == ""
```

Comment:
This method's point is:

1. If *anything* is in the buffer, read a whole line. Because that's our protocol.
2. If *nothing* is in the buffer, exit and return nil.

#1 shouldn't really happen, but I never want to return anything other than a whole line.